r/src/array_from_vector.cpp

```diff
@@ -192,6 +192,15 @@ std::shared_ptr<Array> MakeStructArray(SEXP df, const std::shared_ptr<DataType>&
   return std::make_shared<StructArray>(type, children[0]->length(), children);
 }
 
+std::shared_ptr<Array> MakeListArray(SEXP x, const std::shared_ptr<DataType>& type) {
+  int n = type->num_children();
+  std::vector<std::shared_ptr<Array>> children(n);
+  for (int i = 0; i < n; i++) {
+    children[i] = Array__from_vector(VECTOR_ELT(df, i), type->child(i)->type(), true);
+  }
+  return std::make_shared<ListArray>(type, children[0]->length(), children);
```

Comment:
This isn't a proper construction of a `ListArray` but I can't quite figure out how to make one. I see this in `cpp/src/arrow/array.h` but I'm not sure how to build `value_offsets`:

```
class ARROW_EXPORT ListArray : public BaseListArray<ListType> {
 public:
  explicit ListArray(const std::shared_ptr<ArrayData>& data);

  ListArray(const std::shared_ptr<DataType>& type, int64_t length,
            const std::shared_ptr<Buffer>& value_offsets,
            const std::shared_ptr<Array>& values,
            const std::shared_ptr<Buffer>& null_bitmap = NULLPTR,
            int64_t null_count = kUnknownNullCount, int64_t offset = 0);
```

Reply:
This link is proving helpful:

https://arrow.apache.org/docs/format/Columnar.html#format-columnar